File: sdk/golang/kernel/kernel_vectors_test.go
package kernel

import (
	"encoding/json"
	"os"
	"testing"
)

type threeWayVector struct {
	Name     string `json:"name"`
	Ancestor string `json:"ancestor"`
	A        string `json:"a"`
	B        string `json:"b"`
	Expected string `json:"expected"`
	IsErr    bool   `json:"isErr"`
}

type mergeFieldVector struct {
	Name      string `json:"name"`
	Change    string `json:"change"`
	PolicyRef string `json:"policyRef"`
	Ctx       string `json:"ctx"`
	Expected  string `json:"expected"`
	IsErr     bool   `json:"isErr"`
}

type compareJSONVector struct {
	Name     string `json:"name"`
	A        string `json:"a"`
	B        string `json:"b"`
	Expected string `json:"expected"`
	IsErr    bool   `json:"isErr"`
}

type transformToCifVector struct {
	Name     string `json:"name"`
	Source   string `json:"source"`
	Schema   string `json:"schema"`
	FormatID string `json:"formatId"`
	Expected string `json:"expected"`
	IsErr    bool   `json:"isErr"`
}

type transformFromCifVector struct {
	Name     string `json:"name"`
	Cif      string `json:"cif"`
	Schema   string `json:"schema"`
	FormatID string `json:"formatId"`
	Expected string `json:"expected"`
	IsErr    bool   `json:"isErr"`
}

type resolveVector struct {
	Name      string `json:"name"`
	Ancestor  string `json:"ancestor"`
	Changelog string `json:"changelog"`
	PolicyDoc string `json:"policyDoc"`
	Ctx       string `json:"ctx"`
	Expected  string `json:"expected"`
	IsErr     bool   `json:"isErr"`
}

type mergeBatchVector struct {
	Name      string `json:"name"`
	Changelog string `json:"changelog"`
	PolicyDoc string `json:"policyDoc"`
	Ctx       string `json:"ctx"`
	Expected  string `json:"expected"`
	IsErr     bool   `json:"isErr"`
}

type fuseVector struct {
	Name      string `json:"name"`
	Ancestor  string `json:"ancestor"`
	A         string `json:"a"`
	B         string `json:"b"`
	PolicyDoc string `json:"policyDoc"`
	Ctx       string `json:"ctx"`
	Expected  string `json:"expected"`
	IsErr     bool   `json:"isErr"`
}

type kernelVectors struct {
	ThreeWayDiff     []threeWayVector         `json:"threeWayDiff"`
	MergeField       []mergeFieldVector       `json:"mergeField"`
	CompareJSON      []compareJSONVector      `json:"compareJson"`
	TransformToCif   []transformToCifVector   `json:"transformToCif"`
	MergeBatch       []mergeBatchVector       `json:"mergeBatch"`
	Fuse             []fuseVector             `json:"fuse"`
	TransformFromCif []transformFromCifVector `json:"transformFromCif"`
	Resolve          []resolveVector          `json:"resolve"`
}

// The P4 gate for three_way_diff/merge_field: all vectors, read straight
// from spec/vectors/ (monorepo relative path — the Rust generator is the
// sole producer). Every field is a JSON-encoded string so all runtimes get
// byte-identical input; "expected" is the bare wire string (the ok/err
// envelope is stripped by Kernel.call before it reaches us), so success and
// error cases are both compared as exact strings, never structurally.
func TestKernelVectors(t *testing.T) {
	raw, err := os.ReadFile("../../../spec/vectors/kernel-vectors.json")
	if err != nil {
		t.Fatalf("read vectors: %v", err)
	}
	var vs kernelVectors
	if err := json.Unmarshal(raw, &vs); err != nil {
		t.Fatalf("parse vectors: %v", err)
	}
	if len(vs.ThreeWayDiff) != 17 || len(vs.MergeField) != 29 || len(vs.CompareJSON) != 8 || len(vs.TransformToCif) != 13 || len(vs.MergeBatch) != 5 || len(vs.Fuse) != 7 || len(vs.TransformFromCif) != 10 || len(vs.Resolve) != 7 {
		t.Fatalf("expected 17 threeWayDiff + 29 mergeField + 8 compareJson + 13 transformToCif + 5 mergeBatch + 7 fuse + 10 transformFromCif + 7 resolve vectors, got %d + %d + %d + %d + %d + %d + %d + %d",
			len(vs.ThreeWayDiff), len(vs.MergeField), len(vs.CompareJSON), len(vs.TransformToCif), len(vs.MergeBatch), len(vs.Fuse), len(vs.TransformFromCif), len(vs.Resolve))
	}

	k, ctx := newKernel(t)

	for _, v := range vs.ThreeWayDiff {
		t.Run("threeWayDiff/"+v.Name, func(t *testing.T) {
			out, err := k.ThreeWayDiff(ctx, []byte(v.Ancestor), []byte(v.A), []byte(v.B))
			if v.IsErr {
				if err == nil {
					t.Fatalf("want error, got %s", out)
				}
				if err.Error() != v.Expected {
					t.Errorf("got error %q want %q", err.Error(), v.Expected)
				}
				return
			}
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if string(out) != v.Expected {
				t.Errorf("got %s want %s", out, v.Expected)
			}
		})
	}

	for _, v := range vs.MergeField {
		t.Run("mergeField/"+v.Name, func(t *testing.T) {
			out, err := k.mergeField(ctx, []byte(v.Change), []byte(v.PolicyRef), []byte(v.Ctx))
			if v.IsErr {
				if err == nil {
					t.Fatalf("want error, got %s", out)
				}
				if err.Error() != v.Expected {
					t.Errorf("got error %q want %q", err.Error(), v.Expected)
				}
				return
			}
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if string(out) != v.Expected {
				t.Errorf("got %s want %s", out, v.Expected)
			}
		})
	}

	for _, v := range vs.CompareJSON {
		t.Run("compareJson/"+v.Name, func(t *testing.T) {
			out, err := k.compareJSON(ctx, []byte(v.A), []byte(v.B))
			if v.IsErr {
				if err == nil {
					t.Fatalf("want error, got %s", out)
				}
				if err.Error() != v.Expected {
					t.Errorf("got error %q want %q", err.Error(), v.Expected)
				}
				return
			}
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if string(out) != v.Expected {
				t.Errorf("got %s want %s", out, v.Expected)
			}
		})
	}

	for _, v := range vs.TransformToCif {
		t.Run("transformToCif/"+v.Name, func(t *testing.T) {
			out, err := k.TransformToCIF(ctx, []byte(v.Source), []byte(v.Schema), v.FormatID)
			if v.IsErr {
				if err == nil {
					t.Fatalf("want error, got %s", out)
				}
				if err.Error() != v.Expected {
					t.Errorf("got error %q want %q", err.Error(), v.Expected)
				}
				return
			}
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if string(out) != v.Expected {
				t.Errorf("got %s want %s", out, v.Expected)
			}
		})
	}

	for _, v := range vs.MergeBatch {
		t.Run("mergeBatch/"+v.Name, func(t *testing.T) {
			out, err := k.mergeBatch(ctx, []byte(v.Changelog), []byte(v.PolicyDoc), []byte(v.Ctx))
			if v.IsErr {
				if err == nil {
					t.Fatalf("want error, got %s", out)
				}
				if err.Error() != v.Expected {
					t.Errorf("got error %q want %q", err.Error(), v.Expected)
				}
				return
			}
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if string(out) != v.Expected {
				t.Errorf("got %s want %s", out, v.Expected)
			}
		})
	}

	for _, v := range vs.Fuse {
		t.Run("fuse/"+v.Name, func(t *testing.T) {
			out, err := k.fuse(ctx, []byte(v.Ancestor), []byte(v.A), []byte(v.B), []byte(v.PolicyDoc), []byte(v.Ctx))
			if v.IsErr {
				if err == nil {
					t.Fatalf("want error, got %s", out)
				}
				if err.Error() != v.Expected {
					t.Errorf("got error %q want %q", err.Error(), v.Expected)
				}
				return
			}
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if string(out) != v.Expected {
				t.Errorf("got %s want %s", out, v.Expected)
			}
		})
	}

	for _, v := range vs.TransformFromCif {
		t.Run("transformFromCif/"+v.Name, func(t *testing.T) {
			out, err := k.TransformFromCIF(ctx, []byte(v.Cif), []byte(v.Schema), v.FormatID)
			if v.IsErr {
				if err == nil {
					t.Fatalf("want error, got %s", out)
				}
				if err.Error() != v.Expected {
					t.Errorf("got error %q want %q", err.Error(), v.Expected)
				}
				return
			}
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if string(out) != v.Expected {
				t.Errorf("got %s want %s", out, v.Expected)
			}
		})
	}

	for _, v := range vs.Resolve {
		t.Run("resolve/"+v.Name, func(t *testing.T) {
			out, err := k.Resolve(ctx, []byte(v.Ancestor), []byte(v.Changelog), []byte(v.PolicyDoc), []byte(v.Ctx))
			if v.IsErr {
				if err == nil {
					t.Fatalf("want error, got %s", out)
				}
				if err.Error() != v.Expected {
					t.Errorf("got error %q want %q", err.Error(), v.Expected)
				}
				return
			}
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if string(out) != v.Expected {
				t.Errorf("got %s want %s", out, v.Expected)
			}
		})
	}
}

File: sdk/typescript/tests/unit/kernelVectors.test.ts
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";
import {
  mergePolicyRefSchema,
  policyDocumentSchema,
} from "../../src/application/policy/declaration.js";
import { mergeOutcomeSchema } from "../../src/application/policy/index.js";
import {
  batchResolutionSchema,
  fuseResultSchema,
  kernelCompareJsonRaw,
  kernelFuseRaw,
  kernelMergeBatchRaw,
  kernelMergeFieldRaw,
  kernelThreeWayDiffRaw,
  kernelTransformToCifRaw,
  wireChangelogSchema,
  wireFieldChangeSchema,
} from "../../src/kernel.js";

// Every field below is a JSON-encoded string, not a parsed value: the vector
// file hands each runtime byte-identical input so nothing gets reshaped by
// JSON.stringify/serde/encoding-json on the way in.
type ThreeWayDiffVector = {
  name: string;
  ancestor: string;
  a: string;
  b: string;
  expected: string;
  isErr: boolean;
};

type MergeFieldVector = {
  name: string;
  change: string;
  policyRef: string;
  ctx: string;
  expected: string;
  isErr: boolean;
};

type MergeBatchVector = {
  name: string;
  changelog: string;
  policyDoc: string;
  ctx: string;
  expected: string;
  isErr: boolean;
};

type FuseVector = {
  name: string;
  ancestor: string;
  a: string;
  b: string;
  policyDoc: string;
  ctx: string;
  expected: string;
  isErr: boolean;
};

type CompareJsonVector = {
  name: string;
  a: string;
  b: string;
  expected: string;
  isErr: boolean;
};

type TransformToCifVector = {
  name: string;
  source: string;
  schema: string;
  formatId: string;
  expected: string;
  isErr: boolean;
};

type KernelVectors = {
  threeWayDiff: ThreeWayDiffVector[];
  mergeField: MergeFieldVector[];
  mergeBatch: MergeBatchVector[];
  fuse: FuseVector[];
  compareJson: CompareJsonVector[];
  transformToCif: TransformToCifVector[];
};

const fixtureUrl = new URL(
  "../../../../spec/vectors/kernel-vectors.json",
  import.meta.url,
);
const vectors = JSON.parse(
  readFileSync(fileURLToPath(fixtureUrl), "utf8"),
) as KernelVectors;

const EXPECTED_THREE_WAY_DIFF = 17;
const EXPECTED_MERGE_FIELD = 29;
const EXPECTED_MERGE_BATCH = 5;
const EXPECTED_FUSE = 7;
const EXPECTED_COMPARE_JSON = 8;
const EXPECTED_TRANSFORM_TO_CIF = 13;
if (
  vectors.threeWayDiff.length !== EXPECTED_THREE_WAY_DIFF ||
  vectors.mergeField.length !== EXPECTED_MERGE_FIELD ||
  vectors.mergeBatch.length !== EXPECTED_MERGE_BATCH ||
  vectors.fuse.length !== EXPECTED_FUSE ||
  vectors.compareJson.length !== EXPECTED_COMPARE_JSON ||
  vectors.transformToCif.length !== EXPECTED_TRANSFORM_TO_CIF
) {
  throw new Error(
    `expected ${EXPECTED_THREE_WAY_DIFF} threeWayDiff + ${EXPECTED_MERGE_FIELD} mergeField + ${EXPECTED_MERGE_BATCH} mergeBatch + ${EXPECTED_FUSE} fuse + ${EXPECTED_COMPARE_JSON} compareJson + ${EXPECTED_TRANSFORM_TO_CIF} transformToCif vectors, got ${vectors.threeWayDiff.length} + ${vectors.mergeField.length} + ${vectors.mergeBatch.length} + ${vectors.fuse.length} + ${vectors.compareJson.length} + ${vectors.transformToCif.length}`,
  );
}

/** Runs `fn`, returns the thrown error's message, or fails if it didn't throw. */
function messageOf(fn: () => void): string {
  try {
    fn();
  } catch (e) {
    expect(e).toBeInstanceOf(Error);
    return (e as Error).message;
  }
  throw new Error("expected function to throw");
}

describe("kernel vector conformance", () => {
  describe("three_way_diff", () => {
    for (const v of vectors.threeWayDiff) {
      it(v.name, () => {
        if (v.isErr) {
          expect(messageOf(() => kernelThreeWayDiffRaw(v.ancestor, v.a, v.b))).toBe(
            v.expected,
          );
        } else {
          expect(kernelThreeWayDiffRaw(v.ancestor, v.a, v.b)).toBe(v.expected);
        }
      });
    }
  });

  describe("merge_field", () => {
    for (const v of vectors.mergeField) {
      it(v.name, () => {
        if (v.isErr) {
          expect(
            messageOf(() => kernelMergeFieldRaw(v.change, v.policyRef, v.ctx)),
          ).toBe(v.expected);
        } else {
          expect(kernelMergeFieldRaw(v.change, v.policyRef, v.ctx)).toBe(
            v.expected,
          );
        }
      });
    }
  });

  describe("merge_batch", () => {
    for (const v of vectors.mergeBatch) {
      it(v.name, () => {
        if (v.isErr) {
          expect(
            messageOf(() => kernelMergeBatchRaw(v.changelog, v.policyDoc, v.ctx)),
          ).toBe(v.expected);
        } else {
          expect(kernelMergeBatchRaw(v.changelog, v.policyDoc, v.ctx)).toBe(
            v.expected,
          );
        }
      });
    }
  });

  describe("fuse", () => {
    for (const v of vectors.fuse) {
      it(v.name, () => {
        if (v.isErr) {
          expect(
            messageOf(() => kernelFuseRaw(v.ancestor, v.a, v.b, v.policyDoc, v.ctx)),
          ).toBe(v.expected);
        } else {
          expect(kernelFuseRaw(v.ancestor, v.a, v.b, v.policyDoc, v.ctx)).toBe(
            v.expected,
          );
        }
      });
    }
  });

  describe("compare_json", () => {
    for (const v of vectors.compareJson) {
      it(v.name, () => {
        if (v.isErr) {
          expect(messageOf(() => kernelCompareJsonRaw(v.a, v.b))).toBe(v.expected);
        } else {
          expect(kernelCompareJsonRaw(v.a, v.b)).toBe(v.expected);
        }
      });
    }
  });

  describe("transform_to_cif", () => {
    for (const v of vectors.transformToCif) {
      it(v.name, () => {
        if (v.isErr) {
          expect(
            messageOf(() => kernelTransformToCifRaw(v.source, v.schema, v.formatId)),
          ).toBe(v.expected);
        } else {
          expect(kernelTransformToCifRaw(v.source, v.schema, v.formatId)).toBe(
            v.expected,
          );
        }
      });
    }
  });
});

describe("zod schema drift guard", () => {
  for (const v of vectors.threeWayDiff.filter((v) => !v.isErr)) {
    it(`${v.name}: wireChangelogSchema accepts kernel output`, () => {
      expect(() => wireChangelogSchema.parse(JSON.parse(v.expected))).not.toThrow();
    });
  }
  for (const v of vectors.mergeField.filter((v) => !v.isErr)) {
    it(`${v.name}: schemas accept vector payloads`, () => {
      expect(() => wireFieldChangeSchema.parse(JSON.parse(v.change))).not.toThrow();
      expect(() => mergePolicyRefSchema.parse(JSON.parse(v.policyRef))).not.toThrow();
      expect(() => mergeOutcomeSchema.parse(JSON.parse(v.expected))).not.toThrow();
    });
  }
  for (const v of vectors.mergeBatch.filter((v) => !v.isErr)) {
    it(`${v.name}: schemas accept vector payloads`, () => {
      expect(() => wireChangelogSchema.parse(JSON.parse(v.changelog))).not.toThrow();
      expect(() => policyDocumentSchema.parse(JSON.parse(v.policyDoc))).not.toThrow();
      expect(() => batchResolutionSchema.parse(JSON.parse(v.expected))).not.toThrow();
    });
  }
  for (const v of vectors.fuse.filter((v) => !v.isErr)) {
    it(`${v.name}: schemas accept vector payloads`, () => {
      expect(() => policyDocumentSchema.parse(JSON.parse(v.policyDoc))).not.toThrow();
      expect(() => fuseResultSchema.parse(JSON.parse(v.expected))).not.toThrow();
    });
  }
});
